sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorCoordinator.java

```diff
@@ -0,0 +1,512 @@
+        message.setBody(new AmqpValue(declare));
+        return message;
+    }
+
+    //@Override
+    public Mono<DeliveryState> completeTransaction(AmqpTransaction transaction, boolean isCommit) {
+        if (hasConnected.get()) {
+            return Mono.create(sink -> send(getCompleteTransactionMessage(transaction,
+                isCommit), sink));
+        } else {
+            return RetryUtil.withRetry(
+                handler.getEndpointStates().takeUntil(state -> state == EndpointState.ACTIVE), timeout, retry)
+                .then(Mono.create(sink -> send(getCompleteTransactionMessage(transaction,
+                    isCommit), sink)));
+        }
+    }
+
+    //@Override
+    public Mono<DeliveryState> createTransaction() {
+        if (hasConnected.get()) {
+            return Mono.create(sink -> send(getCreateTransactionMessage(), sink));
+        } else {
+            return RetryUtil.withRetry(
+                handler.getEndpointStates().takeUntil(state -> state == EndpointState.ACTIVE), timeout, retry)
+                .then(Mono.create(sink -> send(getCreateTransactionMessage(), sink)));
+        }
+    }
+
+    /**
+     * Send the AMQP {@link Message}.
+     *
+     * @param message AMQP {@link Message}
+     * @param sink to send back {@link DeliveryState}.
+     */
+    private void send(Message message, MonoSink<DeliveryState> sink) {
+
+        final int payloadSize = messageSerializer.getSize(message);
+        final int allocationSize = payloadSize + MAX_AMQP_HEADER_SIZE_BYTES;
+
+        final byte[] bytes = new byte[allocationSize];
+
+        int encodedSize = message.encode(bytes, 0, allocationSize);
+        send(new RetriableResponseWorkItem(bytes, encodedSize, DeliveryImpl.DEFAULT_MESSAGE_FORMAT,
+            sink, timeout));
+    }
+
+    /**
+     * Add the work item in pending send to be processed on {@link ReactorDispatcher} thread.
+     *
+     * @param workItem to be processed.
+     */
+    private void send(WorkItem workItem) {
+        final String deliveryTag = UUID.randomUUID().toString().replace("-", "");
+
+        synchronized (pendingSendLock) {
+            this.pendingSendsMap.put(deliveryTag, workItem);
+            this.pendingSendsQueue.offer(new ReactorCoordinator.WeightedDeliveryTag(deliveryTag,
+                workItem.hasBeenRetried() ? 1 : 0));
+        }
+
+        this.scheduleWorkOnDispatcher();
+    }
+
+    /**
+     * Invokes work on the Reactor. Should only be called from ReactorDispatcher.invoke()
+     */
+    private void processSendWork() {
+        if (!hasConnected.get()) {
+            logger.warning("Not connected. Not processing send work.");
+            return;
+        }
+        while (hasConnected.get() && sender.getCredit() > 0) {
+            final ReactorCoordinator.WeightedDeliveryTag weightedDelivery;
+            final WorkItem workItem;
+            final String deliveryTag;
+            synchronized (pendingSendLock) {
```

Comment:
This looks like a carbon copy of reactor sender with some modifications. Is there any way to consolidate code?

At its heart, this is basically the reactor sender, that listens to delivered messages.

Reply:
I am merging the logic and removing ReactorCoordinator.